client/src/www/app/outline_server_repository/index.ts

```diff
@@ -58,74 +69,92 @@ interface OutlineServerJson {
+  return repo;
+}
+
 // Maintains a persisted set of servers and liaises with the core.
-export class OutlineServerRepository implements ServerRepository {
-  // Name by which servers are saved to storage.
-  static readonly SERVERS_STORAGE_KEY_V0 = 'servers';
-  static readonly SERVERS_STORAGE_KEY = 'servers_v1';
-  private serverById!: Map<string, OutlineServer>;
-  private lastForgottenServer: OutlineServer | null = null;
+class OutlineServerRepository implements ServerRepository {
+  private lastForgottenServer: ServerEntry | null = null;
+  private serverById = new Map<string, ServerEntry>();
 
   constructor(
     private vpnApi: VpnApi,
     private eventQueue: events.EventQueue,
     private storage: Storage,
     private localize: Localizer
-  ) {
-    console.debug('OutlineServerRepository is initializing');
-    this.loadServers();
-    console.debug('OutlineServerRepository loaded servers');
-    vpnApi.onStatusChange((id: string, status: TunnelStatus) => {
-      console.debug(
-        `OutlineServerRepository received status update for server ${id}: ${status}`
-      );
-      let statusEvent: events.OutlineEvent;
-      switch (status) {
-        case TunnelStatus.CONNECTED:
-          statusEvent = new events.ServerConnected(id);
-          break;
-        case TunnelStatus.DISCONNECTING:
-          statusEvent = new events.ServerDisconnecting(id);
-          break;
-        case TunnelStatus.DISCONNECTED:
-          statusEvent = new events.ServerDisconnected(id);
-          break;
-        case TunnelStatus.RECONNECTING:
-          statusEvent = new events.ServerReconnecting(id);
-          break;
-        default:
-          console.warn(
-            `Received unknown tunnel status ${status} for tunnel ${id}`
-          );
-          return;
-      }
-      eventQueue.enqueue(statusEvent);
-    });
-    console.debug('OutlineServerRepository registered server status callback');
-  }
+  ) {}
 
   getAll() {
-    return Array.from(this.serverById.values());
+    return Array.from(this.serverById.values()).map(e => e.server);
   }
 
   getById(serverId: string) {
-    return this.serverById.get(serverId);
+    return this.serverById.get(serverId)?.server;
   }
 
-  add(accessKey: string) {
+  async add(accessKey: string): Promise<void> {
     const alreadyAddedServer = this.serverFromAccessKey(accessKey);
     if (alreadyAddedServer) {
       throw new errors.ServerAlreadyAdded(alreadyAddedServer);
     }
-    const server = this.createServer(uuidv4(), accessKey, undefined);
+    const server = await this.internalCreateServer(
+      uuidv4(),
+      accessKey,
+      undefined
+    );
 
-    this.serverById.set(server.id, server);
     this.storeServers();
     this.eventQueue.enqueue(new events.ServerAdded(server));
   }
 
   rename(serverId: string, newName: string) {
-    const server = this.serverById.get(serverId);
+    const server = this.serverById.get(serverId)?.server;
```

Comment:
Reuse the existing logic?

```suggestion
    const server = this.getById(serverId);
```